# 🎁 Unpacking & **kwargs — Tags wali Gift-Bag & Star-Openers!

**Folder:** 00_Python → 03_Pythonic_Thinking → 05_Args_Kwargs
**Level:** 🐔 Inter · **Type:** PRACTICAL (💾 products_catalog.csv)

---
## 🎯 Is notebook me kya seekhoge
1. **`**kwargs`** — named extras ka DICT-pack (tags wali gift-bag 🎁)
2. Combined full-signatures: pos → *args → kw-only → **kwargs ek-line flow!
3. Call-side **unpacking**: `f(*list)`, `f(**dict)` (star kholo ⭐!)
4. `functools.partial` teaser (curry-ka-chhota-bhai 🍛)
5. 💾 products → flexible-discount lab + receipt-printer 🧾

**One-liner:**
> `*args` = ANONYMOUS items ki jholi 🛍️; `**kwargs` = NAMED-TAGS wali gift-bag
> 🎁 — definition-side PACK, call-side UNPACK (star double-sided!).

## 🌍 Real-World Hook: Courier-Office ka Form 📮

- Fixed fields: *"name, road, city"* — positional order pe bhare
- Optional fields: *"jo satra extra likhna chahe, likho!"* 🎁 — 
  `landmark=`, `alt_phone=`, `gift_wrap=` — caller apne-hisaab se jod!

`**kwargs` ka kaam: jitne bhi NAMED extras aaye → ek **dictionary** pack karke lo.
Function andar `.get(key, default)` se gracefully handle! 📮

In [1]:
# 🐣 **kwargs — named extras ka dict-pack! 🎁
def profile(**info):
    print(f"type: {type(info).__name__} | keys: {sorted(info)}")
    print(f"  name: {info.get('name', 'N/A')}")
    print(f"  city: {info.get('city', 'unknown')}")

profile(name="Sana", city="Delhi", hobby="chess")
print()
profile()                                            # empty-gift-bag OK!
# Expected OUTPUT:
# type: dict | keys: ['city', 'hobby', 'name']
#   name: Sana
#   city: Delhi
#
# type: dict | keys: []
#   name: N/A
#   city: unknown

type: dict | keys: ['city', 'hobby', 'name']
  name: Sana
  city: Delhi

type: dict | keys: []
  name: N/A
  city: unknown


In [2]:
# 🐥 FULL-SIGNATURE combo — sab ek saath! (pos → *args → kw-only → **kwargs)
def show(order_id, *items, urgent=False, **meta):
    print(f"order#{order_id} items={items}")
    print(f"  urgent={urgent} | meta={meta}")

show(101, "book", "pen", urgent=True, gift="yes", note="call-back")
# Expected OUTPUT:
# order#101 items=('book', 'pen')
#   urgent=True | meta={'gift': 'yes', 'note': 'call-back'}

order#101 items=('book', 'pen')
  urgent=True | meta={'gift': 'yes', 'note': 'call-back'}


In [3]:
# 🐥 Call-side UNPACKING — star se dabba kholo ⭐ (list → args!)
def add3(a, b, c):
    return a + b + c

nums = [10, 20, 30]
print("add3(*nums)     :", add3(*nums), "← list ko * se KHOLA — 3 args ban!")

def welcome(name, city):
    print(f"Welcome {name} from {city}!")

student = {"city": "Pune", "name": "Ravi"}          # order different — dict matches BY NAME!
welcome(**student)                                   # ** se named-arg bomb!
# Expected OUTPUT:
# add3(*nums)     : 60 ← list ko * se KHOLA — 3 args ban!
# Welcome Ravi from Pune!

add3(*nums)     : 60 ← list ko * se KHOLA — 3 args ban!
Welcome Ravi from Pune!


In [4]:
# 🐔 kwargs-length check + kw-detective helpers
def packer(*args, **kwargs):
    return f"anonymous={len(args)} tagged={len(kwargs)}"

print(packer(1, 2, 3, x=1, y=2))
print(packer())                                      # zero-flexibility legit!
# Expected OUTPUT:
# anonymous=3 tagged=2
# anonymous=0 tagged=0

anonymous=3 tagged=2
anonymous=0 tagged=0


In [5]:
# 🐔 partial teaser — curry-ka-chhota-bhai 🍛 (Section 04 OOP ke side-serves!)
from functools import partial

def discount_calc(rate, price):
    return round(price * (1 - rate), 2)

eighteen_off = partial(discount_calc, 0.18)          # rate FIXED, price open!
print("partial      :", eighteen_off(48000), "← Laptop ka 18%-off price")
print("original fnc :", discount_calc(0.18, 48000), "(same result — freeze-then-use convenience!)")
# Expected OUTPUT:
# partial      : 39360.0 ← Laptop ka 18%-off price
# original fnc : 39360.0 (same result — freeze-then-use convenience!)

partial      : 39360.0 ← Laptop ka 18%-off price
original fnc : 39360.0 (same result — freeze-then-use convenience!)


In [6]:
# 💾 REAL DATASET — Products ka Flexible Discount-Lab! 🏷️
import csv

def load_products():
    for path in ["../../00_Resources/datasets/products_catalog.csv",
                 "00_Python/00_Resources/datasets/products_catalog.csv"]:
        try:
            with open(path, newline="", encoding="utf-8") as f:
                return list(csv.DictReader(f))
        except FileNotFoundError:
            continue
    raise FileNotFoundError("products_catalog.csv nahi mila!")

products = load_products()
name2price = {p["product_name"]: float(p["price"]) for p in products}
print("catalog:", len(name2price), "products ready 🏪")

def discounted_list(rate, *prices):
    return [round(p * (1 - rate), 2) for p in prices]

first3 = list(name2price.values())[:3]
print("\n18%-off first-3 :", discounted_list(0.18, *first3))
print("names zipped    :", list(zip(list(name2price)[:3], discounted_list(0.18, *first3))))
# Expected OUTPUT:
# catalog: 30 products ready 🏪
#
# 18%-off first-3 : [39360.0, 12300.0, 1640.0]
# names zipped    : [('Laptop', 39360.0), ('Smartphone', 12300.0), ('Headphones', 1640.0)]

catalog: 30 products ready 🏪

18%-off first-3 : [39360.0, 12300.0, 1640.0]
names zipped    : [('Laptop', 39360.0), ('Smartphone', 12300.0), ('Headphones', 1640.0)]


In [7]:
# 💾 Receipt printer — **kwargs style optional-fields ka showcase! 🧾
def receipt(customer, *items, discount=0.0, gst=0.0, delivery=0.0, **flags):
    subtotal = sum(name2price[i] for i in items)
    after_disc = subtotal * (1 - discount)
    after_gst = after_disc * (1 + gst)
    final = after_gst + delivery
    print(f"  🧾 {customer}'s cart: {len(items)} items")
    for i in items:
        print(f"    - {i}: ₹{name2price[i]:,.0f}")
    print(f"    subtotal ₹{subtotal:,.0f} | -{int(discount*100)}% → ₹{after_disc:,.0f}")
    print(f"    +{int(gst*100)}% GST → ₹{after_gst:,.0f} | +delivery ₹{delivery:,.0f}")
    print(f"    FINAL: ₹{final:,.0f} | flags: {sorted(flags)}")

receipt("Rahul", "Laptop", "Smartphone", "Headphones",
        discount=0.10, gst=0.05, delivery=500, gift_wrap="yes", festival="diwali")
# Expected OUTPUT:
#   🧾 Rahul's cart: 3 items
#     - Laptop: ₹48,000
#     - Smartphone: ₹15,000
#     - Headphones: ₹2,000
#     subtotal ₹65,000 | -10% → ₹58,500
#     +5% GST → ₹61,425 | +delivery ₹500
#     FINAL: ₹61,925 | flags: ['festival', 'gift_wrap']

  🧾 Rahul's cart: 3 items
    - Laptop: ₹48,000
    - Smartphone: ₹15,000
    - Headphones: ₹2,000
    subtotal ₹65,000 | -10% → ₹58,500
    +5% GST → ₹61,425 | +delivery ₹500
    FINAL: ₹61,925 | flags: ['festival', 'gift_wrap']


## ⚠️ 5 Common Mistakes (Unpacking Special)

| # | ❌ Galti | ✅ Sahi |
|---|---|---|
| 1 | `f(*list)` pe short-count args (TypeError!) | Functions ke params = list-length match, ya defaults backup |
| 2 | `**dict` me extra-keys pe classic function reject — unexpected kwarg TypeError | function me `**kwargs` gobble-up laga-lo |
| 3 | kwargs ko caller-ka-data assume karna | kwargs Python ka FRESH dict — caller vars safe, andar play freely! |
| 4 | `f(a=1, *args, a=2)` duplicate kwarg — TypeError! `got multiple values` | name-collisions dedupe karein |
| 5 | Signature me `**kwargs` *args se pehle | ORDER law: pos → *args → kw-only → **kwargs (fixed!) |

> 🎯 **Library-design rule:** public functions me `**kwargs` gobble-slot rakho —
> kal ke naye features keywords se absorb, purane callers unbrekan! 🧙 (pro move)

## 📌 Recap — 30 Second Me

- `**kwargs` = named-extras ka DICT (`.get(key, default)` handling = graceful!).
- Full signature: `def f(pos, *args, kw_only=.., **kwargs)` — flow fixed!
- Call-side: `f(*list)` positional-unpack ⭐, `f(**dict)` name-match unpack (order-free!).
- `partial` = function-freezer 🍛 (some args fix, baaki open) — functoolsmagic.
- 💾 products: discount-lab (18%-off `[39360, 12300, 1640]`) + dynamic receipt 🧾 (Rahul ₹61,925!).

### ⏭️ Next — `task.ipynb` (Section-03 FINAL BOSS!)
7 tasks ⭐→🏆: signature drills se **Smart Bill Maker** tak — flexible-args ka
championship! Jeetoge to Section 03 TROPHY! 🏆🎉